### ISSUE TITLE
The generator does not handle double values very well when they happen to be integers (in checked mode)

### ISSUE THREAD

**Issue by gothamgasworks** [dart-lang/googleapis#48]
Please note I'm demonstrating with the Datastore API, but this applies to all API that has `double` fields.

# Steps to reproduce a problem

1. Set up a project that uses the Cloud Datastore
2. Add an entity to the datastore
   a. Kind: `Foo`
   b. Key identifier: set it to `Custom name`, set the name to `bar`
   c. Add a property
      - name: `field`
      - type: `Floating point number`
      - value: 0

   d. Click Create
3. Set up a Dart project that fetches that entity
4. Run that project in checked mode

# Expected behavior

The checks pass. Whenever there's a `double` field, it should work in checked mode even if the value happens to be an integer.

# Actual behavior

It fails in checked mode, because for example `JSON.decode("0") is int` and `JSON.decode("0") is! double` are both true.

In this example in the `Value` class in the Datastore, there's a line:

```
    if (_json.containsKey("doubleValue")) {
      doubleValue = _json["doubleValue"];
    }
```

If that `doubleValue` field happens to contain an integer, like `{"doubleValue":7}`, it will be an `int` and it will fail in checked mode, the assignment should look something like `doubleValue = _json["doubleValue"]?.toDouble()`

These sources seems to be generated by a more advanced version of https://github.com/google/apis-client-generator/ and that project seems to be pretty much abandoned, so I'm filing here, because this one seems more active (and also, because it's obvious this project uses a modified or more advanced version of that generator).

The fix looks like something like this on https://github.com/google/apis-client-generator/blob/master/src/googleapis/codegen/languages/dart/0.2/templates/_model_content.tmpl replacing the `fromJson` constructor generator:

```
  /** Create new {{ model.className }} from JSON data. */
  {{ model.className }}.fromJson(core.Map json) {
{% for property in model.properties %}
    if (json.containsKey({% literal property.wireName %})) {
{% if property.data_type.arrayOf %}
      {{ property.codeName }} = [];
      json[{% literal property.wireName %}].forEach((item) {
        {% if property.data_type.baseType.builtIn %}
        {{ property.codeName }}.add(item);
        {% else %}
        {% if property.data_type.baseType.code_type == 'core.double' %}
        {{ property.codeName }}.add(new {{ property.data_type.baseType.className }}.fromJson(item?.toDouble()));
        {% else %}
        {{ property.codeName }}.add(new {{ property.data_type.baseType.className }}.fromJson(item));
        {% endif %}
        {% endif %}
      });
{% else %}
      {% if property.data_type.builtIn %}
        {% if property.data_type.code_type == 'core.double' %}
      {{ property.codeName }} = json[{% literal property.wireName %}]?.toDouble();
        {% else %}
      {{ property.codeName }} = json[{% literal property.wireName %}];
        {% endif %}
      {% else %}
      {{ property.codeName }} = new {{ property.data_type.className }}.fromJson(json[{% literal property.wireName %}]);
      {% endif %}
{% endif %}
    }
{% endfor %}
  }
```

**Comment by mkustermann** [comment]
Our [dart-lang/discoveryapis_generator](https://github.com/dart-lang/discoveryapis_generator) should have fixed this recently.

Do you still see your problem?

/cc @jakobr-google

**Comment by gothamgasworks** [comment]
Oops, right, my mistake, should run `pub upgrade` more often :D

**Comment by jakobr-google** [comment]
This was fixed in v 0.50.4, which was released yesterday.

The code in question (`Value.fromJson()` in datastore/v1.dart) now looks like this:
```dart
    ...
    if (_json.containsKey("doubleValue")) {
      doubleValue = _json["doubleValue"].toDouble();
    }
    ...
```

**Comment by jakobr-google** [comment]
Ah, missed it by a second. :)
